Add product query support for Product Summary block

imanish003 opened the pull request:
Adds Product Query support for the atomic Product Summary block

On the client side, when the Product Summary block is used within the product query block, the markup will be rendered on the server side - No javascript related to Product Summary block will be rendered.

Fixes #7334


### Testing
1. Checkout this branch and run npm start
2. Be sure that you have the latest version of the Gutenberg feature plugin installed and activated
3. Create a post/page and add the Product Query block.
4. Add Product Summary block to Post Template of product query block as shown in video below:

https://user-images.githubusercontent.com/16707866/204458088-0e719130-d858-4573-88b5-99d4922513ab.mov

5. Now save the post & confirm that: 
    1. No JS related to the Product Summary block is loaded (search in your network tab for product-summary-frontend.js). As you can see in the screenshot below, product-summary-frontend.js is showing. **This shouldn't be visible in the network tab.**
![image](https://user-images.githubusercontent.com/16707866/204456560-400521b1-3d17-479d-bce9-10f89f3cd9cb.png)
    2. UI of block should be identical in Frontend & Editor.
    3. The settings are reflected correctly (eg: custom style). Try changing custom style from the sidebar.
    4. ADDITIONAL CSS CLASS(ES)(available in advanced toggle in sidebar) should be added to the container div
    5. Summary text styling set using WYSIWYG editor should work in Editor & frontend. By WYSIWYG editor, I mean following editing block of product:
![image](https://user-images.githubusercontent.com/16707866/204457376-3313ca80-9b91-47d6-9802-cafbd8c306ca.png)




* [X] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

imanish003 committed:
Add product query support for Product Summary block

On the client side, when the Product Summary block is used within the product query block, the markup will be rendered on the server side - No javascript related to Product Summary block will be rendered.

imanish003 committed:
Update variable names for clarity & readability

sunyatasattva commented:
Curious here: why skip changelog? I think this is changelog-worthy.

Aljullu commented on src/BlockTypes/ProductSummary.php:
```diff
@@ -51,5 +53,67 @@ protected function get_block_type_supports() {
+		$output  = '<div class="wc-block-components-product-summary ' . $classes_and_styles['classes'] . ' ' . $classname . '"';
+		$output .= $style;
+		$output .= '>';
+		$output .= '<p>' . $summary_content . '</p>';
```
```suggestion
		$output .= '<p>' . wp_kses_post( $summary_content ) . '</p>';
```

Aljullu commented on src/BlockTypes/ProductSummary.php:
```diff
@@ -51,5 +53,67 @@ protected function get_block_type_supports() {
+		$classname          = isset( $attributes['className'] ) ? $attributes['className'] : '';
+		$style              = isset( $classes_and_styles['styles'] ) ? ' style="' . $classes_and_styles['styles'] . '" ' : '';
+
+		$output  = '<div class="wc-block-components-product-summary ' . $classes_and_styles['classes'] . ' ' . $classname . '"';
```
```suggestion
		$output  = '<div class="wc-block-components-product-summary ' . esc_attr( $classes_and_styles['classes'] ) . ' ' . esc_attr( $classname ) . '"';
```

Aljullu commented on src/BlockTypes/ProductSummary.php:
```diff
@@ -51,5 +53,67 @@ protected function get_block_type_supports() {
+		$style              = isset( $classes_and_styles['styles'] ) ? ' style="' . $classes_and_styles['styles'] . '" ' : '';
+
+		$output  = '<div class="wc-block-components-product-summary ' . $classes_and_styles['classes'] . ' ' . $classname . '"';
+		$output .= $style;
```
I would inline the ternary from L91 and add `esc_atr()` around `$classes_and_styles['styles']`.

Aljullu commented on src/BlockTypes/ProductSummary.php:
```diff
@@ -51,5 +53,67 @@ protected function get_block_type_supports() {
+			return null;
+		}
+
+		$summary = wc_trim_string( $summary_content, 150, '...' );
```
[`...`  is the default suffix](https://github.com/woocommerce/woocommerce/blob/410f06b8c8098c9c6be8182c35257745c714308b/plugins/woocommerce/includes/wc-formatting-functions.php#L1082), so I would probably not pass it and rely on the default.
```suggestion
		$summary = wc_trim_string( $summary_content, 150 );
```

imanish003 replied:
Nice catch! I will remove it 🤝

Aljullu reviewed (commented):
Thanks for creating this PR, @imanish003! This is testing well and code overall looks good. However, I think we should escape all values we print in `render()`.

Even when they come from a source we know (ie: `StyleAttributesUtils::get_classes_and_styles_by_attributes()`), we should be escaping everything we print in the HTML. Otherwise, we might introduce a filter in `get_classes_and_styles_by_attributes()` in the future, and a plugin would be able to modify the output and print some malicious code in there.

I saw #7709, #7675 and #7734 are also affected by this. Will you be able to update them (if the PR is open) or create a new PR fixing it (if the original PR has already been merged?). We would need this to be fixed before the release next Monday.

imanish003 committed:
Escape all values in output string

imanish003 commented:
Hi @Aljullu,
Thanks for pointing out the escaping issue 🙏🏻 I have made the changes. Can you please have a look again? Meanwhile, I am going to make the same changes in PRs too 🙂

imanish003 committed:
Merge branch 'trunk' into add/7334-product-summary-block-pq-support

Aljullu commented on src/BlockTypes/ProductSummary.php:
```diff
@@ -51,5 +53,67 @@ protected function get_block_type_supports() {
+		$summary_content = self::get_summary_content( $product );
+
+		if ( ! $summary_content ) {
+			return null;
```
Looking at [this function](https://github.com/WordPress/wordpress-develop/blob/0cb8475c0d07d23893b1d73d755eda5f12024585/src/wp-includes/blocks.php#L980-L990), I think the `render()` function should return a string, so this should probably be:
```suggestion
			return '';
```

imanish003 replied:
Nice catch. I will update it. Thanks for pointing this out 🙂

Aljullu commented on src/BlockTypes/ProductSummary.php:
```diff
@@ -51,5 +53,67 @@ protected function get_block_type_supports() {
+
+		$output  = '<div class="wc-block-components-product-summary '
+			. esc_attr( $classes_and_styles['classes'] ) . ' ' . esc_attr( $classname ) . '"'
+			. isset( $classes_and_styles['styles'] ) ? ' style="' . esc_attr( $classes_and_styles['styles'] ) . '"' : '';
```
The tertiary like this doesn't seem to work well for me (sorry, I know if was my original suggestion). :sweat: 

![imatge](https://user-images.githubusercontent.com/3616980/204796431-3df490a7-331c-43e8-aa1d-93851e4ab736.png)

But I think it you do it as a normal if, it works well, then:

```PHP
if ( isset( $classes_and_styles['styles'] ) ) {
	$output .= ' style="' . esc_attr( $classes_and_styles['styles'] ) . '"';
}
```

imanish003 replied:
Oh, I did test the PR after making the changes, but it looks like my changes weren't reflected correctly while I was doing the testing. Thanks for pointing this out. 🙂

Aljullu reviewed (commented):
Thanks for updating this PR, @imanish003! I had a small issue when the block had some custom styling, but besides that, everything looks good. I commented it inline below.

Aljullu commented:
> Curious here: why skip changelog? I think this is changelog-worthy.

@sunyatasattva I guess it's because the Product Query block hasn't been released yet (until next release), so probably there is no need to mention all its improvements in the changelog until it has been published, no?

sunyatasattva commented:
> > Curious here: why skip changelog? I think this is changelog-worthy.
> 
> @sunyatasattva I guess it's because the Product Query block hasn't been released yet (until next release), so probably there is no need to mention all its improvements in the changelog until it has been published, no?

Yep, @Aljullu, we talked privately with Manish and decided that from next release onwards, these kind of changes should have changelog entries, but for now it's ok like this.

The downside is that, as we release, these won't appear on the list. But I guess our Release Post can cover everything.

imanish003 committed:
Fix custom style not working

More info: https://github.com/woocommerce/woocommerce-blocks/pull/7774#discussion_r1035909243

imanish003 committed:
Merge branch 'trunk' into add/7334-product-summary-block-pq-support

Aljullu reviewed (approved):
This is testing well and code looks good. Thanks @imanish003! :ship:

imanish003 committed:
Merge branch 'trunk' into add/7334-product-summary-block-pq-support
